lib/widgets/message_list.dart

```diff
@@ -635,3 +714,76 @@ final _kMessageTimestampStyle = TextStyle(
+  // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
+  // will be the oldest non-muted unread message, which would
+  // result in muted unreads older than the first unread not
+  // being processed.
+  Anchor anchor = AnchorCode.oldest;
+  int responseCount = 0;
+  int updatedCount = 0;
+  while (true) {
+    final result = await updateMessageFlagsForNarrow(connection,
+      anchor: anchor,
+      // [AnchorCode.oldest] is an anchor ID lower than any valid
+      // message ID; and follow-up requests will have already
+      // processed the anchor ID, so we just want this to be
+      // unconditionally false.
+      includeAnchor: false,
+      // There is an upper limit of 5000 messages per batch
+      // (numBefore + numAfter <= 5000) enforced on the server.
+      // See `update_message_flags_in_narrow` in zerver/views/message_flags.py .
+      // zulip-mobile uses `numAfter` of 5000, but web uses 1000
+      // for more responsive feedback. See zulip@f0d87fcf6.
+      numBefore: 0,
+      numAfter: 1000,
+      narrow: narrow.apiEncode(),
+      op: UpdateMessageFlagsOp.add,
+      flag: MessageFlag.read);
+    if (!context.mounted) return;
+    responseCount++;
+    updatedCount += result.updatedCount;
+
+    if (result.foundNewest) {
+      if (responseCount > 1) {
+        // We previously showed an in-progress [SnackBar], so say we're done.
+        // There may be a backlog of [SnackBar]s accumulated in the queue
+        // so be sure to clear them out here.
+        ScaffoldMessenger.of(context)
+          ..clearSnackBars()
+          ..showSnackBar(SnackBar(behavior: SnackBarBehavior.floating,
+              content: Text(zulipLocalizations.markAsReadComplete(updatedCount))));
+      }
+      return;
+    }
+
+    if (result.lastProcessedId == null) {
+      // No messages were in the range of the request.
+      // This should be impossible given that `foundNewest` was false
+      // (and that our `numAfter` was positive.)
+      await showErrorDialog(context: context,
+        title: zulipLocalizations.errorMarkAsReadFailedTitle,
+        message: zulipLocalizations.errorInvalidResponse);
+      return;
+    }
+    anchor = NumericAnchor(result.lastProcessedId!);
+
+    // The task is taking a while, so tell the user we're working on it.
+    // No need to say how many messages, as the [MarkAsUnread] widget
+    // should follow along.
+    // TODO: Ideally we'd have a progress widget here that showed up based
+    //   on actual time elapsed -- so it could appear before the first
+    //   batch returns, if that takes a while -- and that then stuck
+    //   around continuously until the task ends.  But we don't have an
+    //   off-the-shelf way to wire up such a thing, and marking a giant
+    //   number of messages unread isn't a common enough flow to be worth
+    //   substantial effort on UI polish.  So for now, we use a SnackBar,
+    //   even though they may feel a bit janky.
```

Comment:
```suggestion
    //   around continuously until the task ends.  For now we use
    //   a series of [SnackBar]s, which may feel a bit janky.
```

The justification here is true of marking unread but I think it's less convincing for marking as read, which is what this function is doing.